components/mdns/mdns.c

```diff
@@ -1871,21 +1871,44 @@ static void _mdns_create_answer_from_parsed_packet(mdns_parsed_packet_t *parsed_
     packet->id = parsed_packet->id;
 
     mdns_parsed_question_t *q = parsed_packet->questions;
+    uint32_t out_record_nums = 0;
     while (q) {
         shared = q->type == MDNS_TYPE_PTR || q->type == MDNS_TYPE_SDPTR || !parsed_packet->probe;
         if (q->type == MDNS_TYPE_SRV || q->type == MDNS_TYPE_TXT) {
             mdns_srv_item_t *service = _mdns_get_service_item_instance(q->host, q->service, q->proto, NULL);
             if (service == NULL || !_mdns_create_answer_from_service(packet, service->service, q, shared, send_flush)) {
                 _mdns_free_tx_packet(packet);
                 return;
+            } else {
+                out_record_nums++;
             }
         } else if (q->service && q->proto) {
             mdns_srv_item_t *service = _mdns_server->services;
             while (service) {
                 if (_mdns_service_match_ptr_question(service->service, q)) {
-                    if (!_mdns_create_answer_from_service(packet, service->service, q, shared, send_flush)) {
-                        _mdns_free_tx_packet(packet);
-                        return;
+                    mdns_parsed_record_t *r = parsed_packet->records;
+                    bool is_record_exist = false;
+                    while (r) {
+                        if (service->service->instance && r->host) {
+                            if (_mdns_service_match_instance(service->service, r->host, r->service, r->proto, NULL) && r->ttl > (MDNS_ANSWER_PTR_TTL / 2)) {
+                                is_record_exist = true;
+                                break;
+                            }
+                        } else if (!service->service->instance && !r->host) {
+                            if (_mdns_service_match(service->service, r->service, r->proto, NULL) && r->ttl > (MDNS_ANSWER_PTR_TTL / 2)) {
+                                is_record_exist = true;
+                                break;
+                            }
+                        }
+                        r = r->next;
+                    }
```

Comment:
Maybe I didn't fully understand the change, but could we tell if the `record_exists` earlier when parsing the packet?

Why do we have to alloc and copy the record entries from parsed answers and compare them later?

Reply:
I probably understand now, as at the time of parsing you cannot tell if you're about to create answer?

Reply:
Yes, since the `instance` attribute exists, a single question may correspond to multiple answers. Therefore, when creating a response, it's necessary to check whether there is already a corresponding entry in the answers list of the query. If there is, then no new entry for response should be created (to avoid wasting resources).

It's not reasonable to perform this check during parsing because even if you know that certain entries already exist in the answers of the query during parsing, you still need to create and maintain a `existing entries list`, and compare it with the entries being attempted to be created when creating the response.

Reply:
It's more accurate to say that the current `Allocate space and store answer records` is essentially maintaining a `existing entries list`.

Reply:
> If we want to perform filtering during the parsing, then we should compare the entries existing in the device list with those in the query answer, and then creating and maintaining a list of entries to be created in the response, which will be used when creating the response. However, I think it looks strange and not reasonable.

Oh, okay. I was just curious if we could simplify that and avoid that extra allocation, but you're right that would probably lead to more complications. Thanks for explaining,